`WCProductTagModel` to Room

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: AINFRA-127
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR migrates `WCProductTagModel` to Room.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

Smoke test any feature that uses `WCProductTagModel`.

The tags are under `+ ADD MORE DETAILS` button on the details of the product:

<img width="300" alt="image" src="https://github.com/user-attachments/assets/94a303fc-c88e-49be-a92b-ed8fc0f94935" />



### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img width="1402" alt="Screenshot 2025-05-05 at 13 29 05" src="https://github.com/user-attachments/assets/ca928954-ec03-4e9a-8264-53649943ff11" />

- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Migrate `WCProductTagModel` to Room Entity

Introduce ProductTagsDao, replace SELECT queries with the new Dao.

wzieba committed:
Migrate rest of the product tags database methods to DAO. Make `libs:fluxc-plugin:assembleDebug` successful

wzieba committed:
Add database migration and changes to make `:WooCommerce:assembleWasabiDebug` task pass

wzieba committed:
Introduce `ProductTagsDaoTest` test. Move and adjust first test from ProductSqlUtilsTest there

wzieba committed:
Migrate remaining ProductTagsDao tests.

`:libs:fluxc-plugin:testDebugUnitTest` passes

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/258">22.4</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Run `getProductTagByName` in a non blocking coroutine

In this case, it seems completely safe to use `viewModelScope` right away

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/details/ProductDetailViewModel.kt:
```diff
@@ -2404,7 +2405,7 @@ class ProductDetailViewModel @Inject constructor(
         // verify if the entered tagName exists for the site
         // It so, the tag should be added to the product directly
-        productTagsRepository.getProductTagByName(tagName)?.let {
+        runBlocking { productTagsRepository.getProductTagByName(tagName) }?.let {
```
In previous scenarios and other migrations, I leaned over persisting this behavior (even if it meant blocking a thread), as I didn't want to introduce yet another variable to the migration process. But in this case, I think it's indeed quite safe. I tested it and it looks fine. Suggestion addressed in: befc13cc03f6549578d3a38e938d4ca0692f748d

wzieba committed:
Remove unused import

ParaskP7 committed:
Merge branch 'migrate_product_category_to_room' of github.com:woocommerce/woocommerce-android into migrate_product_tag_to_room

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/WCProductTagModel.kt:
```diff
@@ -1,28 +1,23 @@
+)
+data class WCProductTagModel(
+    val localSiteId: Int = 0,
+    val remoteTagId: Long = 0L, // The unique identifier for this tag on the server
```
Suggestion (💡): As per my [comment](https://github.com/woocommerce/woocommerce-android/pull/14027#issuecomment-2864045241) and [this](https://github.com/woocommerce/woocommerce-android/commit/375d22af9d4d3c014db0c2dead3494aa044e0c19) commit on the `WCProductVariationModel` migration, maybe we should consider refactor this entity too, that is, to utilize `LocalId`/`RemoteId` in favor of `Int`/`Long`.

PS: [Similar](https://github.com/woocommerce/woocommerce-android/pull/13996#discussion_r2084484305)

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/WCProductTagModel.kt:
```diff
@@ -1,28 +1,23 @@
-@RawConstraints(
-        "FOREIGN KEY(LOCAL_SITE_ID) REFERENCES SiteModel(_id) ON DELETE CASCADE",
-        "UNIQUE (REMOTE_TAG_ID, LOCAL_SITE_ID) ON CONFLICT REPLACE"
+//todo: as soon as SiteModel is migrated to Room, add foreign key constraint
```
👍 about this, just wanted you to know I saw that!

PS: [Similar](https://github.com/woocommerce/woocommerce-android/pull/13996#discussion_r2084607683)

ParaskP7 commented on libs/fluxc-plugin/src/test/java/org/wordpress/android/fluxc/persistence/dao/ProductTagsDaoTest.kt:
```diff
@@ -0,0 +1,160 @@
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.fluxc.persistence.ProductSqlUtils
```
Minor (🔍): Consider removing this unused import.

PS: Same applies for this unused `WCProductTagModel` import on `ProductSqlUtilsTest` test suite.

ParaskP7 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/details/ProductDetailViewModel.kt:
```diff
@@ -2404,13 +2404,15 @@ class ProductDetailViewModel @Inject constructor(
-            // a list of newly added tags
-            _addedProductTags.addNewItem(ProductTag(name = tagName))
-            loadProductTags()
+        viewModelScope.launch {
```
👍 about this, just wanted you to know I saw that!

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductTagsDao.kt:
```diff
@@ -0,0 +1,50 @@
+        AND name IN (:tagsNames)
+        """
+    )
+    abstract suspend fun getProductTags(localSiteId: Int, tagsNames: List<String>): List<WCProductTagModel>
```
Minor (🔍): Consider renaming this to `names` so that it matches the signature below, which is using `name` instead of `tagName`. This way it will be easier to understand from the usage, without looking at the implementation, that both `name` and `names` are actually targeting the same column, the `name` column and not different columns, it was confusing for me initially.

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt:
```diff
@@ -938,22 +939,25 @@ class WCProductStore @Inject internal constructor(
-    fun getProductTagsByNames(site: SiteModel, tagNames: List<String>) =
-        ProductSqlUtils.getProductTagsByNames(site.id, tagNames)
+    suspend fun getProductTagsByNames(site: SiteModel, tagNames: List<String>) =
+        productTagsDao.getProductTags(site.localId().value, tagsNames = tagNames)
```
Minor (🔍): Consider using named arguments for both parameters, just like you did below: `productTagsDao.getProductTags(localSiteId = site.localId().value, names = tagNames)`

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt:
```diff
@@ -1748,6 +1753,7 @@ class WCProductStore @Inject internal constructor(
                 response.result != null -> {
                     WooResult(response.result.map { it.product })
                 }
+
```
Minor (🔍): Consider removing this extra line, or maybe better, changing this to:

```kotlin
            when {
                response.isError -> WooResult(response.error)
                response.result != null -> { WooResult(response.result.map { it.product }) }
                else -> WooResult(WooError(WooErrorType.GENERIC_ERROR, UNKNOWN))
            }
```

ParaskP7 reviewed (changes requested):
👋 @wzieba !

I have reviewed and tested this PR as per the instructions, everything works as expected (ish), one more migration (almost) done! ✅ 💪 🚀 

---

Warning (⚠️): This ish above is just because of the fact that during this migration I am not seeing the `WCProductTagModel` table being dropped from the `wp-fluxc` database (just like it was done [here](https://github.com/woocommerce/woocommerce-android/pull/14027/commits/4211e37dc37742a6c59d3da7f22a9c62f921f740) as an example).

`WellSqlConfig.kt`
```kotlin
open class WellSqlConfig : DefaultWellConfig {
    ...
    override fun getDbVersion(): Int {
        return 211
    }

    override fun onUpgrade(...) {
                ...
                210 -> migrateAddOn(ADDON_WOOCOMMERCE, version) {
                    db.execSQL("DROP TABLE IF EXISTS WCProductTagModel")
                }
                ...
    }
    ...
}
```

PS: The `211` db version takes into account that a similar `210` migration will happen as part of #13996 (see ⚠️ on [review](https://github.com/woocommerce/woocommerce-android/pull/13996#pullrequestreview-2832892870)).

-----

I have left one warning (⚠️), one suggestions (💡) and some minor (🔍) comments for you to consider. I am **NOT** going to approve this PR for now, just because of this one warning, instead I am hitting request changes.

---

FYI: As soon as these #14027 and #13966 PRs gets reviewed, tested and merged, my plan is to proceed with these changes myself, unless of course you've already returned from AFK by then. 😊 🤷

wzieba committed:
Merge branch 'trunk' into migrate_product_tag_to_room

wzieba committed:
Use LocalId/RemoteId wrappers for WCProductTagModel

Update DAOs methods signatures to not duplicate information

wzieba committed:
Rename `tagsNames` to `names`

Address: https://github.com/woocommerce/woocommerce-android/pull/14003#discussion_r2084795187

wzieba committed:
Improve readability of `when` expression

wzieba committed:
Drop WCProductTagModel table from WellSql database

It has been migrated to Room

ParaskP7 reviewed (approved):
👋 @wzieba , I did another review on this after those new commits you pushed today and everything LGTM, I also tested the migration/app once more and (obviously) it works as expected. Thanks a lot! 🙇 ❤️ 🚀

It is a ✅ from my side, so I am approving this. I am not sure if you also want someone from the product team to do some more testing on this change before merging. I'll leave that up to you. 👍
